flink-tests/src/test/java/org/apache/flink/test/checkpointing/ZooKeeperHighAvailabilityITCase.java

```diff
@@ -0,0 +1,333 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Integration tests for {@link org.apache.flink.runtime.checkpoint.ZooKeeperCompletedCheckpointStore}.
+ */
+public class ZooKeeperHighAvailabilityITCase extends TestLogger {
+
+	private static final Duration TEST_TIMEOUT = Duration.ofSeconds(10000L);
+
+	private static final int NUM_JMS = 1;
+	private static final int NUM_TMS = 1;
+	private static final int NUM_SLOTS_PER_TM = 1;
+
+	@ClassRule
+	public static final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+	private static File haStorageDir;
+
+	private static TestingServer zkServer;
+
+	private static MiniClusterResource miniClusterResource;
+
+	private static OneShotLatch waitForCheckpointLatch = new OneShotLatch();
+	private static OneShotLatch failInCheckpointLatch = new OneShotLatch();
+	private static OneShotLatch successfulRestoreLatch = new OneShotLatch();
+
+	@BeforeClass
+	public static void setup() throws Exception {
+		zkServer = new TestingServer();
+
+		Configuration config = new Configuration();
+		config.setInteger(ConfigConstants.LOCAL_NUMBER_JOB_MANAGER, NUM_JMS);
+		config.setInteger(ConfigConstants.LOCAL_NUMBER_TASK_MANAGER, NUM_TMS);
+		config.setInteger(TaskManagerOptions.NUM_TASK_SLOTS, NUM_SLOTS_PER_TM);
+
+		haStorageDir = temporaryFolder.newFolder();
+
+		config.setString(HighAvailabilityOptions.HA_STORAGE_PATH, haStorageDir.toString());
+		config.setString(HighAvailabilityOptions.HA_CLUSTER_ID, UUID.randomUUID().toString());
+		config.setString(HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM, zkServer.getConnectString());
+		config.setString(HighAvailabilityOptions.HA_MODE, "zookeeper");
+
+		// we have to manage this manually because we have to create the ZooKeeper server
+		// ahead of this
+		miniClusterResource = new MiniClusterResource(
+			new MiniClusterResource.MiniClusterResourceConfiguration(
+				config,
+				NUM_TMS,
+				NUM_SLOTS_PER_TM));
+
+		miniClusterResource.before();
+	}
+
+	@AfterClass
+	public static void tearDown() throws Exception {
+		miniClusterResource.after();
+
+		zkServer.stop();
+		zkServer.close();
+	}
+
+	/**
+	 * Verify that we don't start a job from scratch if we cannot restore any of the
+	 * CompletedCheckpoints.
+	 *
+	 * <p>Synchronization for the different steps and things we want to observe happens via
+	 * latches in the test method and the methods of {@link CheckpointBlockingFunction}.
+	 *
+	 * <p>The test follows these steps:
+	 * <ol>
+	 *     <li>Start job and block on a latch until we have done some checkpoints
+	 *     <li>Block in the special function
+	 *     <li>Move away the contents of the ZooKeeper HA directory and make it non-writable
```

Comment:
you're never making the directory non-writable; you should also check that this is possible in the first place

Reply:
Ah, I had that in an earlier version but I removed it because it was a bit to fancy because I had a suspicion that wasn't true in the end. I will remove this comment, WDYT?

Reply:
+1 for removing it